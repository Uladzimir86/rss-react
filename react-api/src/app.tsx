import React, { useState, ReactElement } from 'react';
import Table from './table';



const SearchBar = () => {
  const [search, setSearch] = useState('');
  const [checkSubmit, setCheckSubmit] = useState(false);
  const [articles, setArticles] = useState([]);
  const [loading, setLoading] = useState(false);

  function handleSubmit(event) {
    event.preventDefault();
    if(checkSubmit) setCheckSubmit(false)
    if (/\w/.test(search)) {
      setLoading(true);
      fetch(`https://newsapi.org/v2/everything?q=${search}&apiKey=5ada2f2e485b4f11b600736fa619867c`)
        .then(resp => resp.json())
        .then(obj => {
          setLoading(false);
          setArticles(obj.articles);
        })
    } else setCheckSubmit(true);
  }
  return (
    <>
    <form className="searchBar" onSubmit={handleSubmit}>
      {checkSubmit && <span className="error">Please, enter letters or digits!</span>}
      <input type="text" className="searchBar__input" placeholder="Search" value={search} onChange={(e) => {setSearch(e.target.value)}}/>
      <input type="submit" className="searchBar__input searchBar__input_btn" value="Search"  />
      {loading && <span className="error">Loading...</span>}
    </form>
    <Table articles={articles} />
    </>
  )
}
// const Btn = () => {

// }
// const Main = () => {
//   const cardsArray = cards[1].map((item) => (
//     <Card key={item.word} image={item.image} word={item.word} translation={item.translation}/>
//     ))
//   return (
//     <div className="main">
//       { cardsArray }
//     </div>
//   )
// }
const App = (): ReactElement => (
    <div>
      <SearchBar />
    </div>
  )
export default App;