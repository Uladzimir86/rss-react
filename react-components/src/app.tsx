import React, { ReactElement } from 'react';
import cards from './cards';
import Card from './card';

const SearchBar = () => {
  return (
    <div className="searchBar">
      <input type="text" className="searchBar__input" placeholder="Search"></input>
    </div>
  )
}
const Main = () => {
  const cardsArray = cards[1].map((item, index) => {
    return(
    <Card key={item.word + index} image={item.image} word={item.word} translation={item.translation}/>
    )
  })
  console.log(cardsArray)
  return (
    <div className="main">
      { cardsArray }
    </div>
  )
}
const App = (): ReactElement => {
  return (
    <div>
      <SearchBar />
      <Main /> 
    </div>
  )
}
export default App;
