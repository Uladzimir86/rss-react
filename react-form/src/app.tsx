import React, { ReactElement, useState } from 'react';

const Form = () => {
  const [name, setName] = useState('');
  const [surname, setSurname] = useState('');
  return (
    <form className="form">
      <label className="form__label">
        Name:
        <input type="text" className="form__input" value={name} onChange={(e) => setName(e.target.value)} />
      </label>
      <label className="form__label">
        Surname:
        <input type="text" className="form__input" value={surname} onChange={(e) => setSurname(e.target.value)} />
      </label>
    </form>
  )
}

const App = (): ReactElement => <Form />

export default App;
