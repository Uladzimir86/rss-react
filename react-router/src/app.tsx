import React, { ReactElement, useState } from 'react';
import Table from './table';
import Form from './form';


const App = (): ReactElement => {

  const [articles, setArticles] = useState([]);
  const [error, setError] = useState({err: false, errMessage: ''});

  return (
    <>
      <Form setArticles={setArticles} error={error} setError={setError}/>
      {error.err && <span className="error error_table">{error.errMessage}...</span>}
      {!error.err && <Table articles={articles} />}
    </>
  )
}

export default App;