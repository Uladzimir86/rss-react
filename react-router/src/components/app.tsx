import React, { ReactElement, useState } from 'react';
import {
  BrowserRouter as Router, Route, Switch,
  // Switch,
  // Route,
} from "react-router-dom";
import Table from './table';
import Form from './form';
import Header from './header';


const App = (): ReactElement => {

  const [articles, setArticles] = useState([]);
  const [error, setError] = useState({err: false, errMessage: ''});

  return (
    <Router>
      <Header />
        <Switch>
          <Route path="/" exact>
            <Form setArticles={setArticles} error={error} setError={setError}/>
            {error.err && <span className="error error_table">{error.errMessage}...</span>}
            {!error.err && <Table articles={articles} />}
          </Route>
          <Route path="/about">
            <span> Some text</span>
          </Route>
        </Switch>
    </Router>
  )
}

export default App;