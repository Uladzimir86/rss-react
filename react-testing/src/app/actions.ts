interface Error {
  err: boolean, 
  errMessage: string
}
export interface Article {
  author: string,
  content: string,
  publishedAt: string,
  title: string,
  url: string,
  urlToImage: string,
}

export const getArticles = <T>(articles: T[]): {type: string, articles: T[]} => ({type: 'GET_ARTICLES', articles});
export const getErrors = (error: Error): {type: string, error: Error} => ({type: 'GET_ERRORS', error});