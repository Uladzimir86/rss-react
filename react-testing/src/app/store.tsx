import { createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import reducer from './reducer';

export interface RootState {
  articles: [],
  error: {
    err: false, 
    errMessage: '',
  },
}

const store = createStore(
  reducer,
  composeWithDevTools(),
);
export default store;