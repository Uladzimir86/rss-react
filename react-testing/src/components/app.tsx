import React, { ReactElement } from 'react';
import {BrowserRouter as Router, Route} from "react-router-dom";
import { useSelector } from 'react-redux';
import { CSSTransition } from 'react-transition-group';
import Table from './table';
import Form from './form';
import Header from './header';
import Page404 from './404';
import PageDetails from './page-details';
import { RootState } from '../app/store';

const App = (): ReactElement => {

  const {err, errMessage} = useSelector((state: RootState) => state.error);

  return (
    <Router>
      <Header />
      <Route path="/" exact>
        {({ match }) => (
          <CSSTransition
            in={match != null}
            timeout={3000}
            classNames="page"
            unmountOnExit
          >
            <div className="page">
              <Form />
              {err && <span className="error error_table">{errMessage}...</span>}
              {!err && <Table />}
            </div>
          </CSSTransition>
        )}
      </Route>
      <Route path="/about">
        {({ match }) => (
          <CSSTransition
            in={match != null}
            timeout={500}
            classNames="page"
            unmountOnExit
          >
            <div className="page">
              Постановка целей является первым шагом на пути превращения мечты в реальность...
            </div>
          </CSSTransition>
        )}
      </Route>
      <Route path="/details/:id" render={({location}) => {
        const indexPublishedAt = location.search.indexOf('&publishedAt=');
        const title=location.search.slice(7, indexPublishedAt);
        const publishedAt = location.search.slice(indexPublishedAt)
        return (
          <PageDetails title={title} publishedAt={publishedAt}/>
        )
      }} />
      <Route path="/404">
        <Page404 />
      </Route>
    </Router>
  )
}

export default App;