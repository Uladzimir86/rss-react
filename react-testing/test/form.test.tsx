import React from 'react';
import {render, screen, fireEvent, waitFor} from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../src/app/store';
import Form from '../src/components/form';
import {getArticles, getErrors} from '../src/app/actions';

test('should match snapshot', () => {
  const {container} = render(<Provider store={store}><Form /></Provider>);
  expect(container).toMatchSnapshot();
})
test('should show text: ...enter letters...', async () => {
  render(<Provider store={store}><Form /></Provider>);
  const form = document.querySelector('.form');
  if (form) fireEvent.submit(form);
  await waitFor(() => {
    expect(screen.getByText(/enter letters/)).toBeTruthy();
  })
})

test('should not change Page Number', async () => {
  render(<Provider store={store}><Form /></Provider>);
  const btnLeft = screen.getByText('<');
  const btnRight = screen.getByText('>');
  const input = screen.getByTestId('pageNumber') as HTMLInputElement;
  fireEvent.click(btnLeft);
  await waitFor(() => {
    expect(input.value).toBe('1 / 1');
  })
  fireEvent.click(btnRight);
  await waitFor(() => {
    expect(input.value).toBe('1 / 1');
  })
})

test('should call setSearch',  async () => {
  const setSearch: any = jest.spyOn(React, 'useState');
  setSearch.mockImplementation((search: string) => [search, setSearch]);
  render(<Provider store={store}><Form /></Provider>);
  const input = screen.getByPlaceholderText('Search') as HTMLInputElement;
  fireEvent.change(input, {target: {value: 'a'}});
  await waitFor(() => {
    expect(setSearch).toHaveBeenCalled();
   })
})

test('should call ResultsPerPage', () => {
  const setSort: any = jest.spyOn(React, 'useState');
  setSort.mockImplementation((sort: string) => [sort, setSort]);
  render(<Provider store={store}><Form /></Provider>);
  const select = screen.getByRole('combobox');
  fireEvent.change(select);
  expect(setSort).toHaveBeenCalled();
})

test('should call popularity sort', async () => {
  const setSort: any = jest.spyOn(React, 'useState');
  setSort.mockImplementation((sort: string) => [sort, setSort]);
  render(<Provider store={store}><Form /></Provider>);
  const popBtn = screen.getByText('Popularity');
  fireEvent.click(popBtn);
  await waitFor(() => {
    expect(setSort).toHaveBeenCalled();
    expect(setSort).toHaveBeenCalledWith('popularity');
  })
})

test('should call relevancy sort', async () => {
  const setSort: any = jest.spyOn(React, 'useState');
  setSort.mockImplementation((sort: string) => [sort, setSort]);
  render(<Provider store={store}><Form /></Provider>)
  const popBtn = screen.getByText('Relevancy');
  fireEvent.click(popBtn);
  await waitFor(() => {
    expect(setSort).toHaveBeenCalled();
    expect(setSort).toHaveBeenCalledWith('relevancy');
  })
})

test('should call publishedAt sort', async () => {
  const setSort: any = jest.spyOn(React, 'useState');
  setSort.mockImplementation((sort: string) => [sort, setSort]);
  render(<Provider store={store}><Form /></Provider>)
  const popBtn = screen.getByText('Published at');
  fireEvent.click(popBtn);
  await waitFor(() => {
    expect(setSort).toHaveBeenCalled();
    expect(setSort).toHaveBeenCalledWith('publishedAt');
  })
})


test('should call getArticles', () => {
  const mockFn = jest.fn(() => getArticles([]));
  const result = mockFn();
    expect(mockFn).toHaveBeenCalled();
    expect(result.type).toBe('GET_ARTICLES');
})

test('should call getErrors', () => {
  const mockFn = jest.fn(() => getErrors({err: false, errMessage: 'error'}));
  const result = mockFn();
    expect(mockFn).toHaveBeenCalled();
    expect(result.type).toBe('GET_ERRORS');
})
