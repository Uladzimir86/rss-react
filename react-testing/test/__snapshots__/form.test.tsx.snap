// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div>
  <form
    class="form"
  >
    <span
      class="error"
    >
      Please, enter letters or digits!
    </span>
    <div
      class="form__searchBar"
    >
      <input
        class="searchBar__input searchBar1"
        data-testid="searchBarText"
        placeholder="Search"
        type="text"
        value=""
      />
      <input
        class="searchBar__input searchBar__input_btn"
        type="submit"
        value="Search"
      />
    </div>
    <div
      class="form__sort"
    >
      <button
        class="btn-sort"
        type="submit"
      >
        Relevancy
      </button>
      <button
        class="btn-sort"
        type="submit"
      >
        Popularity
      </button>
      <button
        class="btn-sort"
        type="submit"
      >
        Published at
      </button>
    </div>
    <div
      class="form__pagination"
    >
      <button
        class="btn-pagination"
        id="btnLeft"
        type="submit"
      >
        &lt;
      </button>
      <input
        data-testid="pageNumber"
        disabled=""
        type="text"
        value="1 / 1"
      />
      <button
        class="btn-pagination"
        id="btnRight"
        type="submit"
      >
        &gt;
      </button>
      <label
        class="form__label"
        for="resultsPerPage"
      >
        Results per page:
        <select
          class="form__input"
          name="resultsPerPage"
        >
          <option
            value="10"
          >
            10
          </option>
          <option
            value="20"
          >
            20
          </option>
          <option
            value="40"
          >
            40
          </option>
          <option
            value="80"
          >
            80
          </option>
        </select>
      </label>
    </div>
  </form>
</div>
`;
